### PR TITLE
Feature/crdt or set

### PR DESCRIPTION
**Pull request by pshemass**
first implementation of of ORSet and Vector Clock

### CODE REVIEW

File: src/main/scala/scalaz/ziokeeper/VectorClock.scala [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  implicit class VectorClockOps[A](val v: VectorClock[A]) extends AnyVal {
+    def time(mid: A): Long = v.value.getOrElse(mid, 0L)
+
+    def increment(mid: A, value: Long): VectorClock[A] =
+      VectorClock(v.value.updated(mid, time(mid) + value))
+  }
+
+  sealed trait CompareVectorClock
+
+  case object TheSame extends CompareVectorClock
```
Comment (rabbitonweb) [inline src/main/scala/scalaz/ziokeeper/VectorClock.scala]: What's the difference between `TheSame` and `Concurrent`?

File: src/main/scala/scalaz/ziokeeper/crdt/ORSet.scala [hunk 2]
```diff
@@ -0,0 +1,33 @@
+package scalaz.ziokeeper.crdt
+
+import scalaz.ziokeeper.VectorClock
+
+case class ORSet[A, B](values: Vector[ORSet.VersionedEntry[A, B]])
+
+object ORSet {
+  implicit class ORSetOps[A, B](val oRSet: ORSet[A, B]) extends AnyVal {
```
Comment (mijicd) [inline src/main/scala/scalaz/ziokeeper/crdt/ORSet.scala]: Maybe it makes sense to have these as methods in `ORSet`.

File: src/main/scala/scalaz/ziokeeper/VectorClock.scala [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  def empty[A] = VectorClock[A](value = Map.empty)
+
+  implicit class VectorClockOps[A](val v: VectorClock[A]) extends AnyVal {
+    def time(mid: A): Long = v.value.getOrElse(mid, 0L)
+
+    def increment(mid: A, value: Long): VectorClock[A] =
+      VectorClock(v.value.updated(mid, time(mid) + value))
+  }
+
+  sealed trait CompareVectorClock
```
[the remark this answers is not in the document]
Comment (pshemass) [reply src/main/scala/scalaz/ziokeeper/VectorClock.scala]: I was trying to avoid ambiguity with scala.Ordering which is not the same and I need to implement Ordering anyway but I need to read one more paper to figure out what to do when I we have concurrent modifications of the same elements.

So do you have better name for that?

File: src/main/scala/scalaz/ziokeeper/VectorClock.scala [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+  case object TheSame extends CompareVectorClock
+
+  case object Before extends CompareVectorClock
+
+  case object After extends CompareVectorClock
+
+  case object Concurrent extends CompareVectorClock
+
+  def compare[A](left: VectorClock[A], right: VectorClock[A]): CompareVectorClock = {
```
[the remark this answers is not in the document]
Comment (pshemass) [reply src/main/scala/scalaz/ziokeeper/VectorClock.scala]: sure

File: src/test/scala/scalaz/ziokeeper/crdt/ORSetSpec.scala [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+import org.specs2.mutable.Specification
+import scalaz.ziokeeper.VectorClock
+
+class ORSetSpec extends Specification {
+  override def is =
+    s2"""
+    This specification is for ORSet
+
+    ORSet should
+      add and remove element $e1
```
[the remark this answers is not in the document]
Comment (pshemass) [reply src/test/scala/scalaz/ziokeeper/crdt/ORSetSpec.scala]: what do you mean?

File: src/main/scala/scalaz/ziokeeper/VectorClock.scala (hunk 1 above)
Comment (pshemass) [reply src/main/scala/scalaz/ziokeeper/VectorClock.scala]: the idea behind vector clock is pretty simple 
- increment you local counter whatever you perform actions(e.g. add, remove, modify) as well as when you receive messages from other members of the cluster.

when we have cluster with three member with ids 111, 222, 333 vector clock initially will look like this on all nodes.
(111 -> 0, 222 -> 0, 333 -> 0) 

at this point the are all **TheSame**

imagine that 111 and 222 perform add action 
111 = (111 -> 1, 222 -> 0, 333 -> 0)
222 = (111 -> 0, 222 -> 1, 333 -> 0)
333 =  (111 -> 0, 222 -> 0, 333 -> 0)
111 and 222 are concurrent because you cannot say which happen first.

let me know if that clear enough.

File: src/test/scala/scalaz/ziokeeper/crdt/ORSetSpec.scala [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    val set     = ORSet.empty[Int, Int]
+    val value   = 100
+    //when
+    val withValue              = set.add(value, vc)
+    val (removedVcs, newORset) = withValue.localRemove(value)
+    val removed                = withValue.remove(value, removedVcs)
+    val notRemoved             = withValue.remove(200, removedVcs)
+    val notRemoved2            = withValue.remove(value, Vector(VectorClock.empty[Int].increment(localId, 2)))
+    //then
+    withValue.values must_=== Vector(ORSet.VersionedEntry(100, vc))
```
[the remark this answers is not in the document]
Comment (pshemass) [reply src/test/scala/scalaz/ziokeeper/crdt/ORSetSpec.scala]: how `containTheSameElementsAs` is different from this?

File: src/main/scala/scalaz/ziokeeper/crdt/ORSet.scala (hunk 2 above)
Comment (pshemass) [reply src/main/scala/scalaz/ziokeeper/crdt/ORSet.scala]: sure

File: src/test/scala/scalaz/ziokeeper/crdt/ORSetSpec.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (mijicd) [reply src/test/scala/scalaz/ziokeeper/crdt/ORSetSpec.scala]: Just wandering whether we could express this as a property-based test. Bad choice of wording from my side, I should have mentioned law :).

File: src/test/scala/scalaz/ziokeeper/crdt/ORSetSpec.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (mijicd) [reply src/test/scala/scalaz/ziokeeper/crdt/ORSetSpec.scala]: They're the same, I referred to this line wanting to point out we should stick to one style. The reason I suggested `containTheSameElementsAs` were these:

```scala
    removedVcs must haveSize(1)
    removedVcs.head must_=== vc
```

While all of these checks are in place, I'd rather avoid doing both checks (`must_==` is also fine). Also, `must haveSize(0)` can be replaced with `must be empty`.

File: src/main/scala/scalaz/ziokeeper/VectorClock.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (mijicd) [reply src/main/scala/scalaz/ziokeeper/VectorClock.scala]: I see. Not at the moment, let's leave it as it is, and change once we find the more suitable name.

File: src/test/scala/scalaz/ziokeeper/crdt/ORSetSpec.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (pshemass) [reply src/test/scala/scalaz/ziokeeper/crdt/ORSetSpec.scala]: you are right this is test is hard to follow. I broke it into couple of them. let me know what do you think.

File: src/test/scala/scalaz/ziokeeper/crdt/ORSetSpec.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (pshemass) [reply src/test/scala/scalaz/ziokeeper/crdt/ORSetSpec.scala]: I was thinking about scalaCheck for that but i was too lazy...

**Comment by rabbitonweb** [comment]
> Looks good to me! @jdegoes, @rabbitonweb any remarks that should be fixed before merging?

Give me 24h for my remarks

File: src/main/scala/scalaz/ziokeeper/VectorClock.scala [hunk 7]
```diff
@@ -0,0 +1,63 @@
+package scalaz.ziokeeper
+
+import scala.annotation.tailrec
+import VectorClock._
+case class VectorClock[A](value: Map[A, Long]) extends AnyVal {
```
Comment (jdegoes) [inline src/main/scala/scalaz/ziokeeper/VectorClock.scala]: This is great for now, ultimately, `Map` won't be an efficient way to store or represent a vector clock. 👍

File: src/main/scala/scalaz/ziokeeper/VectorClock.scala [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+  def time(mid: A): Long = value.getOrElse(mid, 0L)
+
+  def increment(mid: A, value: Long): VectorClock[A] =
```
Comment (jdegoes) [inline src/main/scala/scalaz/ziokeeper/VectorClock.scala]: This vector clock will only ever be incremented by `1`. We will maybe need to "set" certain components for other nodes.

File: src/main/scala/scalaz/ziokeeper/VectorClock.scala [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        case _ if leftBigger && rightBigger => (leftBigger, rightBigger)
+        case Nil                            => (leftBigger, rightBigger)
+        case h :: t =>
+          val diff = leftWithDefault(h) - rightWithDefault(h)
+          loop(t, leftBigger || diff > 0, rightBigger || diff < 0)
+      }
+
+    val keys = leftWithDefault.keySet.union(rightWithDefault.keySet)
+
+    loop(keys.toList, leftBigger = false, rightBigger = false) match {
```
Comment (jdegoes) [inline src/main/scala/scalaz/ziokeeper/VectorClock.scala]: This looks potentially non-deterministic (due to the use of `Set`), which will not satisfy laws for equality. 

Consider sorting by `A`, for which you'll need an ordering on `A`. This will be expensive and speaks to the fact that we ultimately need a different representation for `VectorClock`, namely, two arrays (one for `A`, sorted in order, and one for `Long`, with the same ordering). But for now the inefficiency can be ignored.

File: src/main/scala/scalaz/ziokeeper/VectorClock.scala [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+      case (false, false) => TheSame
+      case (true, true)   => Concurrent
+      case (true, false)  => After
+      case (false, true)  => Before
+    }
+  }
+
+  def merge(right: VectorClock[A]): VectorClock[A] = {
```
Comment (jdegoes) [inline src/main/scala/scalaz/ziokeeper/VectorClock.scala]: Consider extracting out `mergeWith(right)((x: Long, y: Long) => Long)`, then you can implement this as `mergeWith(right)(_ max _)`.

File: src/main/scala/scalaz/ziokeeper/VectorClock.scala [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+            acc + (key -> math.max(leftWithDefault(key), rightWithDefault(key)))
+        }
+    )
+  }
+}
+
+object VectorClock {
+  def empty[A] = VectorClock[A](value = Map.empty)
+
+  sealed trait CompareVectorClock
```
Comment (jdegoes) [inline src/main/scala/scalaz/ziokeeper/VectorClock.scala]: This would more classically be called `Ordering` with `EQ`, `GT`, `LT`. I am not sure what `Concurrent` is.

File: src/main/scala/scalaz/ziokeeper/crdt/ORSet.scala [hunk 12]
```diff
@@ -0,0 +1,31 @@
+package scalaz.ziokeeper.crdt
```
Comment (jdegoes) [inline src/main/scala/scalaz/ziokeeper/crdt/ORSet.scala]: The new root package will probably be:

```scala
package dev.zio.keeper
```

Or maybe just:

```scala
package zio.keeper
```

File: src/main/scala/scalaz/ziokeeper/crdt/ORSet.scala [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+import scalaz.ziokeeper.VectorClock
+
+case class ORSet[A, B](values: Vector[ORSet.VersionedEntry[A, B]]) {
```
Comment (jdegoes) [inline src/main/scala/scalaz/ziokeeper/crdt/ORSet.scala]: This is gonna be super slow. But I think that's OK. We should try to make it correct first, and we can optimize later. 👍

File: src/main/scala/scalaz/ziokeeper/crdt/ORSet.scala (hunk 12 above)
Comment (mijicd) [reply src/main/scala/scalaz/ziokeeper/crdt/ORSet.scala]: Shall we migrate the project to new organization?

Comment (jdegoes) [reply src/main/scala/scalaz/ziokeeper/crdt/ORSet.scala]: Yes, no rush on that, but to avoid cluttering `scalaz` we should get all the active ZIO projects to the new org (and delete the remainder). 👍

Comment (mijicd) [reply src/main/scala/scalaz/ziokeeper/crdt/ORSet.scala]: In that case, let's polish this PR, and I'll take care of the migration and packages cleanup.

File: src/main/scala/scalaz/ziokeeper/VectorClock.scala (hunk 9 above)
Comment (pshemass) [reply src/main/scala/scalaz/ziokeeper/VectorClock.scala]: you are right this is not most efficient implementation but I try to make this correct in the first place then optimize.

File: src/main/scala/scalaz/ziokeeper/VectorClock.scala (hunk 10 above)
Comment (pshemass) [reply src/main/scala/scalaz/ziokeeper/VectorClock.scala]: make sense

File: src/main/scala/scalaz/ziokeeper/VectorClock.scala (hunk 11 above)
Comment (pshemass) [reply src/main/scala/scalaz/ziokeeper/VectorClock.scala]: but this is not exactly ordering in the classic `EQ`, `GT`, `LT` sense. 

this is explanation I gave to @rabbitonweb for a question about different between `TheSame` and `Concurrent` I hope that will help you understand `Concurrent`

> the idea behind vector clock is pretty simple

> increment you local counter whatever you perform actions(e.g. add, remove, modify) as well as when you receive and send messages from/to other members of the cluster.
> when we have cluster with three member with ids 111, 222, 333 vector clock initially will look like this on all nodes.
> (111 -> 0, 222 -> 0, 333 -> 0)
> 
> at this point the are all TheSame
> 
> imagine that 111 and 222 perform add action
> 111 = (111 -> 1, 222 -> 0, 333 -> 0)
> 222 = (111 -> 0, 222 -> 1, 333 -> 0)
> 333 = (111 -> 0, 222 -> 0, 333 -> 0)
> 111 and 222 are Concurrent because you cannot say which happen first.

I need to implement Ordering/PartialOrdering to be able to sort values in ORSet so I named this differently(pretty badly though) to avoid confusion.

The problem that I struggle now is how to make sure that result will be deterministic when we have `Concurrent` update on the same key.

File: src/main/scala/scalaz/ziokeeper/VectorClock.scala (hunk 10 above)
Comment (pshemass) [reply src/main/scala/scalaz/ziokeeper/VectorClock.scala]: done

File: src/main/scala/scalaz/ziokeeper/VectorClock.scala (hunk 11 above)
Comment (pshemass) [reply src/main/scala/scalaz/ziokeeper/VectorClock.scala]: renamed to `VectorClockOrdering`

File: src/main/scala/scalaz/ziokeeper/VectorClock.scala (hunk 8 above)
Comment (pshemass) [reply src/main/scala/scalaz/ziokeeper/VectorClock.scala]: changed to `def increament(mid: A)`. we don't need set because merge should be enough.
